port_forwarding_test.go

```diff
@@ -22,101 +22,95 @@ const (
-				server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-					fmt.Fprint(w, expectedResponse)
-				}))
-				freePorts[i] = server.Listener.Addr().(*net.TCPAddr).Port
-				tt.Cleanup(func() {
-					server.Close()
-				})
-			}
-
-			req := testcontainers.GenericContainerRequest{
-				// hostAccessPorts {
-				ContainerRequest: testcontainers.ContainerRequest{
-					Image:           "alpine:3.17",
-					HostAccessPorts: freePorts,
-					Cmd:             []string{"top"},
-				},
-				// }
-				Started: true,
-			}
-
-			var nw *testcontainers.DockerNetwork
-			if tc.hasNetwork {
-				var err error
-				nw, err = network.New(context.Background())
-				require.NoError(tt, err)
-				testcontainers.CleanupNetwork(t, nw)
-
-				req.Networks = []string{nw.Name}
-				req.NetworkAliases = map[string][]string{nw.Name: {"myalpine"}}
-			}
-
-			ctx := context.Background()
-			if !tc.hasHostAccess {
-				var cancel context.CancelFunc
-				ctx, cancel = context.WithTimeout(ctx, 10*time.Second)
-				defer cancel()
-			}
-
-			c, err := testcontainers.GenericContainer(ctx, req)
-			testcontainers.CleanupContainer(t, c)
-			require.NoError(tt, err)
-
-			if tc.hasHostAccess {
-				// create a container that has host access, which will
-				// automatically forward the port to the container
-				assertContainerHasHostAccess(tt, c, freePorts...)
-			} else {
-				// force cancellation because of timeout
-				time.Sleep(11 * time.Second)
-
-				assertContainerHasNoHostAccess(tt, c, freePorts...)
-			}
-		})
+	var nw *testcontainers.DockerNetwork
+	if hasNetwork {
+		var err error
+		nw, err = network.New(context.Background())
+		require.NoError(t, err)
+		testcontainers.CleanupNetwork(t, nw)
+
+		req.Networks = []string{nw.Name}
+		req.NetworkAliases = map[string][]string{nw.Name: {"myalpine"}}
+	}
+
+	ctx := context.Background()
+	if !hasHostAccess {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, 10*time.Second)
+		defer cancel()
+	}
+
+	c, err := testcontainers.GenericContainer(ctx, req)
+	testcontainers.CleanupContainer(t, c)
+	require.NoError(t, err)
+
+	if hasHostAccess {
+		// create a container that has host access, which will
+		// automatically forward the port to the container
+		assertContainerHasHostAccess(t, c, freePorts...)
+	} else {
+		// force cancellation because of timeout
+		time.Sleep(11 * time.Second)
+
+		assertContainerHasNoHostAccess(t, c, freePorts...)
 	}
 }
 
 func httpRequest(t *testing.T, c testcontainers.Container, port int) (int, string) {
 	t.Helper()
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
 	// wgetHostInternal {
 	code, reader, err := c.Exec(
-		context.Background(),
-		[]string{"wget", "-q", "-O", "-", fmt.Sprintf("http://%s:%d", testcontainers.HostInternal, port)},
+		ctx,
+		[]string{"wget", "-q", "-O", "-", "-T", "2", fmt.Sprintf("http://%s:%d", testcontainers.HostInternal, port)},
```

Comment:
question: we are setting 2s timeout just because it's higher than the 1s context cancellation. Is my assumption correct?

Reply:
The context passed into `Exec` doesn't limit the time a process has to run, could be a bug, so we need the timeout on `wget` to ensure it doesn't block forever. Using a larger value ensures that we'll get a context cancellation error.